src/Microsoft.AspNetCore.Mvc.RazorPages/PageModel.cs

```diff
@@ -37,11 +44,66 @@ public PageArgumentBinder Binder
             }
         }
 
-        public Page Page => PageContext.Page;
+        /// <summary>
+        /// Gets or sets the <see cref="IUrlHelper"/>.
+        /// </summary>
+        public IUrlHelper Url
+        {
+            get
+            {
+                if (_urlHelper == null)
+                {
+                    var factory = HttpContext?.RequestServices?.GetRequiredService<IUrlHelperFactory>();
+                    _urlHelper = factory?.GetUrlHelper(PageContext);
+                }
 
+                return _urlHelper;
+            }
+            set
+            {
+                if (value == null)
+                {
+                    throw new ArgumentNullException(nameof(value));
+                }
+
+                _urlHelper = value;
+            }
+        }
+
+        /// <summary>
+        /// Gets the <see cref="RazorPages.Page"/> instance this model belongs to.
+        /// </summary>
+        public Page Page => PageContext?.Page;
+
+        /// <summary>
+        /// Gets the <see cref="RazorPages.PageContext"/>.
+        /// </summary>
         [PageContext]
```

Comment:
The base type for models is optional, so if you wanted to create a custom base type, there needs to be a way to inject the context in to it. `PageModel` uses the same mechanism. The page factory on the other hand assumes you derive from the `Page` type and sets it. https://github.com/aspnet/Mvc/blob/dev/src/Microsoft.AspNetCore.Mvc.RazorPages/Infrastructure/DefaultPageFactory.cs#L62-L63